core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/common/nearby/RandomNearbyIterator.java

```diff
@@ -33,7 +35,7 @@ public RandomNearbyIterator(MemoizingSupply<NearbyDistanceMatrix<Object, Object>
 
     @Override
     public boolean hasNext() {
-        return replayingIterator.hasNext() && nearbySize > 0;
+        return (origin != null || replayingIterator.hasNext()) && nearbySize > 0;
```

Comment:
Would you mind explaining why you needed to touch nearby selection?
I'd expect ruin moves to use nearby as is, without changes - did you perhaps fix a bug?
If so, please file the bug and submit the fix in a different PR, with its own tests.

Reply:
_MimicReplayingEntitySelector_ doesn't allow us to call next more than once, because after first one _recordingAlreadyReturned_ is set to true. It looks intentional, but in the same time prevent us from using nearby selection for a bigger group of entities than 2. The documentation of both _nearbyEntityIterators_ however states that "origin here will be constant unless _next()_ on the original recording iterator is called first". Then there is really no need to call next on the _replayingEntitySelector_ as long as it doesn't change in the meantime. Therefore the workaround not to call it in this case, which also overcomes the limitation of _MimicReplayingEntitySelector_ . Does it look like a bug for you? Or maybe I am using the iterators in a wrong way?